libz-rs-sys/src/gz.rs

```diff
@@ -521,89 +520,127 @@ unsafe fn deallocate_cstr(s: *mut c_char) {
-    // deallocation cannot safely happen until the references are out of scope. Therefore
-    // the free_state call handles the deallocation using the raw pointer to the state.
-    unsafe { free_state(file.cast::<GzState>()) };
-
-    ret
+    match mode {
+        GzMode::GZ_READ => unsafe { gzclose_r(file) },
+        _ => unsafe { gzclose_w(file) },
+    }
 }
 
-// Close a gzip file that was opened for reading.
-//
-// # Returns
-//
-// * Z_OK if `state` has no outstanding error and the file is closed successfully.
-// * A Z_ error code if the `state` is null or the file close operation fails.
-//
-// # Safety
-//
-// `state` must have been properly initialized, e.g. by [`gzopen_help`].
-unsafe fn gzclose_r(state: &mut GzState) -> c_int {
-    if state.mode != GzMode::GZ_READ {
-        return Z_STREAM_ERROR;
-    }
+/// Close a gzip file that was opened for reading.
+///
+/// # Returns
+///
+/// * Z_OK if `state` has no outstanding error and the file is closed successfully.
+/// * A Z_ error code if the `state` is null or the file close operation fails.
+///
+/// # Safety
+///
+/// `file` must be one of the following:
+/// - A file handle must have been obtained from a function in this library, such as [`gzopen`].
+/// - A null pointer.
+#[cfg_attr(feature = "export-symbols", export_name = crate::prefix!(gzclose_r))]
+pub unsafe extern "C-unwind" fn gzclose_r(file: gzFile) -> c_int {
+    let ret = {
+        // Note: `state` lives in this nested block because it must go out of scope before
+        // we can safely delete the allocation later in this function.
+        let Some(state) = (unsafe { file.cast::<GzState>().as_mut() }) else {
+            return Z_STREAM_ERROR;
+        };
 
-    // Free buffers and close the file
-    if state.in_size != 0 {
-        // Safety: state.stream was properly initialized as a z_stream in gzopen_help.
-        unsafe { inflateEnd(&mut state.stream as *mut z_stream) };
-    }
-    let err = match state.err {
-        Z_BUF_ERROR => Z_BUF_ERROR,
-        _ => Z_OK,
+        // Check that we're reading.
+        if state.mode != GzMode::GZ_READ {
+            return Z_STREAM_ERROR;
+        }
+
+        // Process any buffered input.
+        if state.in_size != 0 {
+            // Safety: state.stream was properly initialized as a z_stream in gzopen_help.
+            unsafe { inflateEnd(&mut state.stream as *mut z_stream) };
+        }
+
+        let err = match state.err {
+            Z_BUF_ERROR => Z_BUF_ERROR,
+            _ => Z_OK,
+        };
+
+        match unsafe { libc::close(state.fd) } {
+            0 => err,
+            _ => Z_ERRNO,
+        }
     };
-    match unsafe { libc::close(state.fd) } {
-        0 => err,
-        _ => Z_ERRNO,
-    }
+
+    // Now that the `state` reference has gone out of scope, we can safely
+    // delete the underlying allocation.
+    unsafe { free_state(file.cast::<GzState>()) };
```

Comment:
just to be sure: miri actively complains if you don't put the above in a block? In some cases rust can see that a reference is no longer in use and then more things are allowed.

Reply:
Miri doesn't complain, so I'll un-nest the block and add a safety comment on the `free_state` to document that the reference can't be used beyond that point.